monitor/sth_fetcher.go

```diff
@@ -105,28 +187,135 @@ func (f *sthFetcher) stop() {
+//    a consistency proof between the two STHs from the log.
+// 3. "failed-to-verify-proof" - the monitor returned a proof of consistency
+//    between the two STHs that did not verify.
+// When the monitor fetches a consistency proof from the log it publishes the
+// latency of the operation to the `sth_proof_latency` prometheus histogram.
+func (f *sthFetcher) verifySTHConsistency(firstSTH, secondSTH *ct.SignedTreeHead) {
+	if firstSTH == nil || secondSTH == nil {
+		f.logErrorf("firstSTH or secondSTH was nil")
+		return
+	}
+
+	firstTreeSize := firstSTH.TreeSize
+	firstHash := firstSTH.SHA256RootHash[:]
+
+	secondTreeSize := secondSTH.TreeSize
+	secondHash := secondSTH.SHA256RootHash[:]
+
+	// If the two STH's have equal tree sizes then we expect the SHA256RootHash to
+	// match. If it doesn't match there is no need to check the consistency proofs
+	// because the log is definitely inconsistent. In this case publish an
+	// increment to the `sthInconsistencies` stat
+	if firstTreeSize == secondTreeSize && !bytes.Equal(firstHash, secondHash) {
+		errorLabels := prometheus.Labels{"uri": f.logURI, "type": "equal-treesize-inequal-hash"}
+		f.stats.sthInconsistencies.With(errorLabels).Inc()
+		f.logErrorf("first STH and second STH have same tree size (%d) "+
+			"but different tree hashes. first.SHA256RootHash: %x "+
+			"second.SHA256RootHash: %x",
+			firstTreeSize,
+			firstHash,
+			secondHash)
+		return
+	} else if firstTreeSize == secondTreeSize {
+		// If the two STH's have equal tree sizes and equal SHA256RootHashes there
+		// isn't anything to do. We need STH's from two different tree states to
+		// verify consistency
+		f.logf("first STH and second STH have same SHA256RootHash (%x) "+
+			"and tree size (%d). No consistency proof required",
+			firstHash, firstTreeSize)
+		return
+	}
+
+	// proofDescription is used in log messages to describe the proof being
+	// fetched/verified
+	proofDescription := fmt.Sprintf(
+		"from treesize %d (hash %x) to treesize %d (hash %x)",
+		firstTreeSize, firstHash, secondTreeSize, secondHash)
+
+	// Fetch the consistency proof between the two tree sizes from the log.
+	// Observe the latency of this operation using the `sthProofLatency` stat. If
+	// the operation fails, consider it an inconsistency and publish an increment
+	// to the `sthInconsistencies` stat.
+	ctx, cancel := context.WithTimeout(context.Background(), f.sthTimeout)
+	defer cancel()
+	start := f.clk.Now()
+	f.logf("Getting consistency proof %s", proofDescription)
+	consistencyProof, err := f.client.GetSTHConsistency(ctx, firstTreeSize, secondTreeSize)
+	elapsed := f.clk.Since(start)
+	labels := prometheus.Labels{"uri": f.logURI}
+	f.stats.sthProofLatency.With(labels).Observe(elapsed.Seconds())
+	if err != nil {
+		errorLabels := prometheus.Labels{"uri": f.logURI, "type": "failed-to-get-proof"}
+		f.stats.sthInconsistencies.With(errorLabels).Inc()
+		f.logErrorf("failed to get consistency proof %s: %q : %#v",
```

Comment:
We'll probably get this somewhat often, due to outages or slowness. Under the current design, we'll drop the STH and move on. Let's add both full STHs to the error log, so we can try verifying consistency after any outage resolves.